File: codingMinutes/backtracking/permutationsOfString.cpp
#include<bits/stdc++.h>
using namespace std;

void permute(string &s, int start, int end) {
    if(start == end) {
        cout << s << endl;
        return;
    } else {
        for(int i = start;i<=end; i++) {
            swap(s[start],s[i]);
            permute(s,start+1,end);
            swap(s[i],s[start]);
            return;
        }
    }
}

int main() {
    string input;
    cin >> input;
    int n = input.size()-1;
    permute(input,0,n);
    return 0;
}

File: codingMinutes/backtracking/permutationsOfString.js
const arr = ['a', 'b', 'c', 'd', 'e', 'f']

const swap = (arr, i, j) => {
    const temp = arr[i]
    arr[i] = arr[j];
    arr[j] = temp;
}

const permute = (arr, i) => {
    if (i == arr.length) {
        console.log(arr)
        return
    }
    for (let j = i; j < arr.length; j++) {
        swap(arr, i, j);
        permute(arr, i + 1);
        swap(arr, i, j)
    }
}

permute(arr, 0);